### PR TITLE
[pf CLI] List all package tools in current environment

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
image

List all package tools in current environment.
```
(prompt_github) D:\Project\github_promptflow\promptflow>pf tool list
{
    "package": {
        "promptflow.tools.embedding.embedding": {
            "name": "Embedding",
            "description": "Use Open AI's embedding model to create an embedding vector representing the input text.",
            "type": "python",
            "module": "promptflow.tools.embedding",
            "function": "embedding",
            "inputs": {
                "connection": {
                    "type": [
                        "AzureOpenAIConnection",
                        "OpenAIConnection"
                    ]
                },
                "deployment_name": {
                    "type": [
                        "string"
                    ],
                    "enabled_by": "connection",
                    "enabled_by_type": [
                        "AzureOpenAIConnection"
                    ],
                    "capabilities": {
                        "completion": false,
                        "chat_completion": false,
                        "embeddings": true
                    },
                    "model_list": [
                        "text-embedding-ada-002",
                        "text-search-ada-doc-001",
                        "text-search-ada-query-001"
                    ]
                },
                "model": {
                    "type": [
                        "string"
                    ],
                    "enabled_by": "connection",
                    "enabled_by_type": [
                        "OpenAIConnection"
                    ],
                    "enum": [
                        "text-embedding-ada-002",
                        "text-search-ada-doc-001",
                        "text-search-ada-query-001"
                    ]
                },
                "input": {
                    "type": [
                        "string"
                    ]
                }
            },
            "package": "promptflow-tools",
            "package_version": "0.1.0b6"
        },
        "promptflow.tools.serpapi.SerpAPI.search": {
            "name": "Serp API",
            "description": "Use Serp API to obtain search results from a specific search engine.",
            "inputs": {
                "connection": {
                    "type": [
                        "SerpConnection"
                    ]
                },
                "engine": {
                    "default": "google",
                    "enum": [
                        "google",
                        "bing"
                    ],
                    "type": [
                        "string"
                    ]
                },
                "location": {
                    "default": "",
                    "type": [
                        "string"
                    ]
                },
                "num": {
                    "default": "10",
                    "type": [
                        "int"
                    ]
                },
                "query": {
                    "type": [
                        "string"
                    ]
                },
                "safe": {
                    "default": "off",
                    "enum": [
                        "active",
                        "off"
                    ],
                    "type": [
                        "string"
                    ]
                }
            },
            "type": "python",
            "module": "promptflow.tools.serpapi",
            "class_name": "SerpAPI",
            "function": "search",
            "package": "promptflow-tools",
            "package_version": "0.1.0b6"
        }
    },
    "code": {}
}
```
List package tools and code tools in flow:
```
(prompt_github) D:\Project\github_promptflow\promptflow>pf tool list --flow chat_flow

{
    "package": {
        "promptflow.tools.embedding.embedding": {
            "name": "Embedding",
            "description": "Use Open AI's embedding model to create an embedding vector representing the input text.",
            "type": "python",
            "module": "promptflow.tools.embedding",
            "function": "embedding",
            "inputs": {
                "connection": {
                    "type": [
                        "AzureOpenAIConnection",
                        "OpenAIConnection"
                    ]
                },
                "deployment_name": {
                    "type": [
                        "string"
                    ],
                    "enabled_by": "connection",
                    "enabled_by_type": [
                        "AzureOpenAIConnection"
                    ],
                    "capabilities": {
                        "completion": false,
                        "chat_completion": false,
                        "embeddings": true
                    },
                    "model_list": [
                        "text-embedding-ada-002",
                        "text-search-ada-doc-001",
                        "text-search-ada-query-001"
                    ]
                },
                "model": {
                    "type": [
                        "string"
                    ],
                    "enabled_by": "connection",
                    "enabled_by_type": [
                        "OpenAIConnection"
                    ],
                    "enum": [
                        "text-embedding-ada-002",
                        "text-search-ada-doc-001",
                        "text-search-ada-query-001"
                    ]
                },
                "input": {
                    "type": [
                        "string"
                    ]
                }
            },
            "package": "promptflow-tools",
            "package_version": "0.1.0b6"
        },
        "promptflow.tools.serpapi.SerpAPI.search": {
            "name": "Serp API",
            "description": "Use Serp API to obtain search results from a specific search engine.",
            "inputs": {
                "connection": {
                    "type": [
                        "SerpConnection"
                    ]
                },
                "engine": {
                    "default": "google",
                    "enum": [
                        "google",
                        "bing"
                    ],
                    "type": [
                        "string"
                    ]
                },
                "location": {
                    "default": "",
                    "type": [
                        "string"
                    ]
                },
                "num": {
                    "default": "10",
                    "type": [
                        "int"
                    ]
                },
                "query": {
                    "type": [
                        "string"
                    ]
                },
                "safe": {
                    "default": "off",
                    "enum": [
                        "active",
                        "off"
                    ],
                    "type": [
                        "string"
                    ]
                }
            },
            "type": "python",
            "module": "promptflow.tools.serpapi",
            "class_name": "SerpAPI",
            "function": "search",
            "package": "promptflow-tools",
            "package_version": "0.1.0b6"
        }
    },
    "code": {
        "chat.jinja2": {
            "type": "llm",
            "inputs": {
                "chat_history": {
                    "type": [
                        "string"
                    ]
                },
                "question": {
                    "type": [
                        "string"
                    ]
                }
            },
            "source": "chat.jinja2"
        },
        "show_answer.py": {
            "type": "python",
            "inputs": {
                "chat_answer": {
                    "type": [
                        "string"
                    ]
                }
            },
            "source": "show_answer.py",
            "function": "show_answer"
        }
    }
}

```

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py [hunk 1]
```diff
@@ -80,3 +81,9 @@ def _parse_tool_from_function(f, initialize_inputs=None):
             function=f.__name__,
             module=f.__module__,
         )
+
+    @staticmethod
+    def list():
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_tool_operations.py:86]: add docstring for this fucntion, esp. doc the return data obj

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py [hunk 2]
```diff
@@ -80,3 +81,9 @@ def _parse_tool_from_function(f, initialize_inputs=None):
             function=f.__name__,
             module=f.__module__,
         )
+
+    @staticmethod
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_tool_operations.py:85]: why is this static method?

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/operations/_tool_operations.py]: Add the docstring to list func

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/operations/_tool_operations.py]: remove static

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 3]
```diff
@@ -65,9 +68,38 @@ def add_parser_init_tool(subparsers):
     )
 
 
+def add_parser_list_tool(subparsers):
+    """Add tool list parser to the pf tool subparsers."""
+    epilog = """
+Examples:
+
+# List all package tool in the environment:
```
Comment (D-W-) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:76]: Is this a public API? may update existing doc for CLI in next PR. https://microsoft.github.io/promptflow/reference/pf-command-reference.html#pf-run-create

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 4]
```diff
@@ -65,9 +68,38 @@ def add_parser_init_tool(subparsers):
...
+pf tool list
+# List all package tool and code tool in the flow:
+pf tool list --flow flow-path
+"""  # noqa: E501
+    add_param_flow = lambda parser: parser.add_argument(  # noqa: E731
+        "--flow", type=str, help="the flow directory"
```
Comment (D-W-) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:82]: Do we also support specifying flow dag file?

**Review by D-W- — APPROVED** [review]

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:82]: Yes, it also support specifying flow dag file. But now all flow description in CLI points to flow folder.

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:76]: Update the cli doc

File: docs/reference/pf-command-reference.md [hunk 5]
```diff
@@ -654,4 +654,74 @@ pf run restore --name
...
+
+List all tools in the environment.
+
+```bash
+pf tool list [--flow]
+```
+
+#### Examples
+
+List all package tool in the environment
```
Comment (wangchao1230) [inline docs/reference/pf-command-reference.md:711]: missing ending .

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py [hunk 6]
```diff
@@ -80,3 +83,26 @@ def _parse_tool_from_function(f, initialize_inputs=None):
             function=f.__name__,
             module=f.__module__,
         )
+
+    def list(
+        self,
+        flow: Union[str, PathLike],
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_tool_operations.py:89]: shall this have default value None?

File: docs/reference/pf-command-reference.md (hunk 5 above)
Comment (lalala123123) [reply docs/reference/pf-command-reference.md]: fix it

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/operations/_tool_operations.py]: set default value to flow

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 7]
```diff
@@ -95,3 +127,10 @@ def init_tool(args):
     ToolPackageGenerator(tool_name=args.tool).generate_to_file(script_code_path / f"{args.tool}.py")
     InitGenerator().generate_to_file(script_code_path / "__init__.py")
     print(f'Done. Created the tool "{args.tool}" in {script_code_path.resolve()}.')
+
+
+@exception_handler("Tool list")
+def list_tool(args):
+    pf_client = PFClient()
+    package_tools = pf_client._tools.list(args.flow)
```
Comment (0mza987) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:135]: Is tools operations public, like use it as  `pf.tools.list`?

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py [hunk 8]
```diff
@@ -1344,6 +1345,40 @@ def test_tool_init(self, capsys):
             outerr = capsys.readouterr()
             assert f"The tool name {invalid_tool_name} is a invalid identifier." in outerr.out
 
+    def test_tool_list(self, capsys):
+        with tempfile.TemporaryDirectory():
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1349]: Is this tmep folder being used?

Comment (lalala123123) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1349]: remove useless code

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 7 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:135]: tool operation is currently private and will not be exposed to user

**Review by 0mza987 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**